text/0000-ppv.md

```diff
@@ -0,0 +1,1387 @@
+            a[0] * shuffle!(b[i], [0,0,0,0]) + 
+            a[1] * shuffle!(b[i], [1,1,1,1]) +
+            a[2] * shuffle!(b[i], [2,2,2,2]) +
+            a[3] * shuffle!(b[i], [3,3,3,3]);
+    }
+    r
+}
+```
+  
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+  
+## Vector types
+
+The vector types are named according to the following scheme:
+
+> {element_type}{lane_width}x{number_of_lanes}
+
+where the following element types are introduced by this RFC:
+
+* `i`: signed integer
+* `u`: unsigned integer
+* `f`: float
+* `m`: mask
+
+So that `u16x8` reads "a SIMD vector of eight packed 16-bit wide unsigned
+integers". The width of a vector can be computed by multiplying the
+`{lane_width}` times the `{number_of_lanes}`. For `u16x8`, 16 x 8 = 128, so
+this vector type is 128 bits wide.
+
+This RFC proposes adding all vector types with sizes in range [16, 256] bit to
+the `std::simd` module, that is:
+
+* 16-bit wide vectors: `i8x2`, `u8x2`, `m8x2`
+* 32-bit wide vectors: `i8x4`, `u8x4`, `m8x4`, `i16x2`, `u16x2`,  `m16x2`
+* 64-bit wide vectors: `i8x8`, `u8x8`, `m8x8`, `i16x4`, `u16x4`, `m16x4`,
+  `i32x2`, `u32x2`, `f32x2`, `m32x2`
+* 128-bit wide vectors: `i8x16`, u8x16`, `m8x16`, `i16x8`, `u16x8`, `m16x8`,
+  `i32x4`, `u32x4`, `f32x4`, `m32x4`, `i64x2`, `u64x2`, `f64x2`, `m64x2`
+* 256-bit wide vectors: `i8x32`, u8x32`, m8x32`, i16x16`, u16x16`, m16x16`,
+  i32x8`, u32x8`, f32x8`, m32x8`, i64x4`, u64x4`, f64x4`, `m64x4`
+
+Note that this list is not comprehensive. In particular:
+
+* half-float `f16xN`: these vectors are supported in many architectures (ARM,
+  AArch64, PowerPC64, RISC-V, MIPS, ...) but their support is blocked on Rust
+  half-float support.
+* AVX-512 vector types, not only 512-bit wide vector types, but also `m1xN`
+  vector masks. These are blocked on `std::arch` AVX-512 support.
+* other vector types: x86, AArch64, PowerPC and others include types like
+  `i64x1`, `u64x1`, `f64x1`, `m64x1`, `i128x1`, `u128x1`, `m128x1`, ... These
+  can be always added later as the need for these arises, potentially in combination with the stabilization of the `std::arch` intrinsics for those
+  architectures.
+
+## API of portable packed SIMD vector types
+
+### Traits overview
+
+All vector types implement the following traits:
+
+* `Copy`
+* `Clone`
+* `Default`: zero-initializes the vector.
+* `Debug`: formats the vector as `({}, {}, ...)`.
+* `PartialEq<Self>`: performs a lane-wise comparison between two vectors and
+  returns `true` if all lanes compare `true`. It is equivalent to
+  `a.eq(b).all()`.
+* `PartialOrd<Self>`: compares two vectors lexicographically.
+* `From/Into` lossless casts between vectors with the same number of lanes.
+
+All signed integer, unsigned integer, and floating point vector types implement
+the following traits:
+
+* `{Add,Sub,Mul,Div,Rem}<RHS=Self,Output=Self>`,
+  `{Add,Sub,Mul,Div,Rem}Assign<RHS=Self>`: vertical (lane-wise) arithmetic
+  operations.
```

Comment:
It's probably worth describing how behavior of these operators is different (or not different) from the same operators for non-vector arithmetic types - e.g. wrapping vs panicking on overflow.
(Same applies to bit-shift traits.)

Reply:
Thanks, I'm on this. @sunfish also requested to know the behavior of `Div`/`Rem` when dividing by zero, so I'll add that as well.

Reply:
From looking at rustc's implementation this is the behavior that they seem to have:

* on integer vectors overflowing results in the correct mathematical result modulo 2^n (see exceptions below)
* on floating point vectors overflowing has the same behavior as overflowing a f32 or f64 (+-INFINITY IIRC).

* on integer vectors we directly map division and rem to LLVM's `{s,u}{div,rem}` which state that if any element is divided by zero or the operation overflows  these operations result in undefined behavior.

So this is what we are currently doing. Now the question: can we do better? 

* For `Add`, `Sub`, `Mul`, `Shl`, `Shr`: we could use `nuw`/`nsw` to generate a poison value and get the same behavior than for the scalar integers. We could then expose wrapping arithmetic via `wrapping_{add,sub,mul,shl,shr}`.

* For `Div` and `Rem`: we could expose these as `unsafe fn {div, rem}(self, o: Self) -> Self` methods with a precondition on division by zero, and provide an implementation in the traits that checks that no element is zero before performing the operation, panicking if this isn't true. This is "expensive": one vector comparison + one horizontal reduction, and whether it is better than not providing these operations at all is debatable. Another alternative might be to not expose these operations.

Reply:
@petrochenkov I've gone ahead and updated the RFC which what I consider the "sane"  arithmetic and shift operator semantics:

* the arithmetic and shift traits behave like the operators for the scalars. That is, for integers they panic on overflow and division by zero, and for floats they produce the same results as the scalar.

* the integer vectors also implement a set of `wrapping_{add,sub,...}` methods that perform modulo 2^n arithmetic, just like the scalar integers do

* the integer vectors also implement two `unsafe wrapping_{div,rem}_unchecked` methods that perform no checks.

Reply:
What's sane here is a hard question.
On one hand, the only reason for SIMD's existence is improving performance and `uNxM + uNxM` should ideally turn into a single hardware SIMD addition instruction, but checked operations can't be lowered into a single instruction.
On the other hand, consistency with scalar arithmetic, error detection and everything...

Reply:
I think that any SIMD proposal that does not mesh with the rest of the language is not going to be "good enough". So what I meant with "sane" was the same defaults than the rest of the std library, while at the same time offering easy-to-reach more efficient operations with slightly different semantics. While the `std` library puts these typically in `core::intrinsics`, the portable SIMD vector types have these as inherent methods, making them more "first class". 

IMO the SIMD std facilities have two jobs: protecting users from easy mistakes, and allowing them to write efficient code. These two jobs are at constant tension. 

I don't think that unsafe operations should become the default, but this does not mean that they should be hard to reach for either.

Reply:
@hsivonen I've added that overflow checking only happens in debug mode.

Reply:
FWIW "release/debug" is not the right distinction for overflow checks. They're a separate switch (`-C overflow-checks=[on|off]`) that can be flipped independently of debug info, optimization level, etc. -- it's just by default tied to `debug_assertions` which in turn is by default tied to release/debug.